A self-hosted API Explorer as a new extension

bajtos opened the pull request:
In this pull request, I am picking up the best parts from #1664 and #1976 and proposing an implementation of a self-hosted API explorer build on the features that are already available. The internal design is not as pretty & clean as it could be, but it works OOTB (as long as the user configures a custom url path, see below).

I expect that the internal details of the explorer component will evolve over time as we keep adding more extension points to our REST layer, but that's way beyond this pull request.

The missing piece: how to disable the built-in redirect to `explorer.loopback.io`. ~~I'll be opening a new pull request for that (hopefully today).~~ See _feat(rest): add config option to disable API Explorer redirects #2016_.

Close #1664, close #1976, close #1960.

This pull request is one of the tasks for _[API Explorer] Self-hosted API Explorer #559_.

@strongloop/loopback-maintainers please review

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

bajtos commented:
@strongloop/loopback-maintainers How should we name the extension providing self-hosted API Explorer?

Currently, I am using `@loopback/explorer`, even though the component is using "ApiExplorer" name everywhere. Should we rename it to `@loopback/api-explorer`?

Also, do we want to start some kind of convention for LB extensions? In LB 3.x, extension packages have names starting with `loopback-component-`, e.g. `loopback-component-explorer`. Should we do the same in LB4 and call the new package `@loopback/extension-api-explorer`? Personally, I find such name awfully long and prefer shorter `@loopback/api-explorer`.

When deciding between `explorer` and `api-explorer` (or even something else): Are we expecting other kinds of explorers to be available in LB4 apps in the future? I believe GraphQL has an interactive explorer/client. I think we should use a naming scheme that will make it clear which extension is exploring which protocol. Both `explorer` and `api-explorer` seem too generic to me, I am proposing `rest-explorer` (to go with `@loopback/rest` component) or perhaps `openapi-explorer`.

Thoughts?

raymondfeng commented on packages/explorer/src/explorer.controller.ts:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  indexRedirect() {
+    this.response.redirect(301, this.request.url + '/');
```
See https://stackoverflow.com/questions/13442377/redirect-all-trailing-slashes-globally-in-express.

bajtos replied:
I don't think this is a viable solution. Please note that we don't want to add a trailing slash to every endpoint (e.g. `/api/users/123` -> `/api/users/123/`), only the paths where we serve HTML pages.

raymondfeng commented:
@bajtos CI is failing.

bajtos commented:
> CI is failing.

I know, this is waiting for _feat(rest): add config option to disable API Explorer redirects #2016_ to be landed first.

raymondfeng commented:
The proposal combines the ability of controllers and static assets to serve API explorer with server-side templating. I think it's acceptable first step. LGTM.

bajtos commented:
Re: package name - thank you @jannyHou @marioestradarosa @dhmlau for the feedback. I am going to use `@loopback/rest-explorer` going forward.

bajtos commented on packages/explorer/src/explorer.controller.ts:
```diff
@@ -0,0 +1,63 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+// TODO(bajtos) Allow users to customize the template
```
I'll add a short note to the README. 

FWIW, to customize the UI look, I think users may need to tweak more than just the index html file, e.g. provide a custom css file instead of the default `swagger-ui.css`.

Let me open a new GH issue to keep track of this missing feature: https://github.com/strongloop/loopback-next/issues/2023

bajtos commented on packages/explorer/src/explorer.keys.ts:
```diff
@@ -0,0 +1,16 @@
+ */
+export namespace ApiExplorerBindings {
+  export const CONFIG = BindingKey.create<ApiExplorerConfig>(
+    'explorer.config',
```
+1

I am using `rest-explorer.config` now.

bajtos commented:
I have rebased the patch on the latest master and addressed all remaining issues & todos. The pull request is ready for final review. I'll squash the commit history before landing.

To make the review easier, I left the first three commits unchanged and added a bunch of new small commits (one for each logical step, kind of):
chore: rename explorer to rest-explorer
- fix(rest): allow users to disable Explorer redirects after RestServer… 
- fix: disable built-in swagger-ui redirects when a self-hosted compone… 
- docs: describe rest-explorer component
- fix: apidocs for rest-explorer
- docs: add a mention of hard-coded visual style to README
- feat(cli): enable self-hosted REST API explorer

@raymondfeng @jannyHou @marioestradarosa LGTY?

bajtos commented:
The documentation changes require us to land https://github.com/strongloop/loopback.io/pull/767, I think that other PR should be landed first.

bajtos commented:
> The documentation changes require us to land strongloop/loopback.io#767, I think that other PR should be landed first.

Actually, I found a better solution - see https://github.com/strongloop/loopback-next/pull/2025

bajtos commented:
> continuous-integration/travis-ci/pr — The Travis CI build failed

```
✖   scope must be one of [
    benchmark, docs, example-hello-world, example-log-extension, example-rpc-server, 
    example-soap-calculator, example-todo-list, example-todo, authentication, boot, build, 
    cli, context, core, http-caching-proxy, http-server, metadata, openapi-spec-builder,
    openapi-v3-types, openapi-v3, repository-json-schema, repository, rest-explorer, rest, 
    service-proxy, testlab, sandbox-example
] [scope-enum]
✖   found 1 problems, 0 warnings
```

Please ignore this build failure. It's just the commit linter using a wrong version of package list to verify scope. This error will get fixed when I clean up the commit history.

bajtos commented:
```
Liquid Exception: Could not locate the included file
  'readmes/loopback-next/packages/rest-explorer/README.md' in any of
  ["/home/travis/build/strongloop/loopback-next/sandbox/loopback.io/pages/en/lb4"].
  Ensure it exists in one of those directories and, if it is a symlink, does not point 
  outside your site source. in /_layouts/readme.html
```

The failure of `verify-docs` build step will be fixed once https://github.com/strongloop/loopback-next/pull/2025 is landed and this pull request is rebased on the new master.

b-admike commented on packages/cli/generators/app/templates/src/application.ts.ejs:
```diff
@@ -1,5 +1,9 @@
+import {
+  RestExplorerBindings,
+  RestExplorerComponent,
+} from '@loopback/rest-explorer';
```
should we add `this.component(RestExplorerComponent)` in the generated application class as well?

b-admike replied:
This is resolved in a later commit.

b-admike commented on packages/rest/src/rest.server.ts:
```diff
@@ -244,16 +244,9 @@ export class RestServer extends Context implements Server, HttpServerLike {
-      return;
-    }
-
     const explorerPaths = ['/swagger-ui', '/explorer'];
```
Should `explorerPaths` take into account custom paths the user configures with our `RestExplorer` component?

bajtos replied:
I am not opposed to that, but I think it's out of scope of this pull request.

b-admike commented on packages/rest-explorer/README.md:
````diff
@@ -35,6 +35,14 @@ this.bind(RestExplorerBindings.CONFIG).to({
 });
 ```
 
+_NOTE: The Explorer UI's visual style is not customizable yet. Our recommend
````
nitpick: `recommend` -> `recommended`

b-admike commented on docs/site/Self-hosted-REST-API-Explorer.md:
```diff
@@ -0,0 +1,13 @@
+summary:
+  A tutorial on using `@loopback/rest-explorer` to add a self-hosted REST API
+  Explorer
+---
```
Are you planning to add docs to this page in this PR?

bajtos replied:
This page will automatically include the contents of `packages/rest-explorer/README.md`, see https://loopback.io/doc/en/lb4/Creating-decorators.html as an example of a page based on README contents.

bajtos commented on packages/explorer/test/acceptance/explorer.acceptance.ts:
```diff
@@ -0,0 +1,119 @@
+  let request: Client;
+
+  beforeEach(function resetAppAndRequestFromPreviousTest() {
+    (app as unknown) = undefined;
```
This is a hack allowing me to set `app` to `undefined`. From the point of view of TypeScript, `app` is not allowed to contain `undefined` values, only an application instance.

I guess I'll need to rethink this part.

raymondfeng commented on docs/site/Server.md:
```diff
@@ -139,6 +139,15 @@ const app = new RestApplication({
+working internet connection is required to explore the API. As a recommended
+alternative, LoopBack comes with an extension that provides a self-hosted
+Explorer UI. Please refer to
+[Self-hosted REST API Explorer](./Self-hosted-rest-api-explorer.md) for more
```
The file name `./Self-hosted-rest-api-explorer.md` is different from `docs/site/Self-hosted-REST-API-Explorer.md` by case.

bajtos replied:
Good catch 👍

cloudwheels commented:
@bajtos when will this extension be released? Thanks

bajtos commented:
I believe `@loopback/rest-explorer@1.1.1` has been already published on November 17th, see https://github.com/strongloop/loopback-next/commit/e0fe37d086e4fbb1cf8df23731e7b7637dd550c1
